### PR TITLE
[EH] Add exnref type back

### PR DESCRIPTION
**Pull request by aheejin**
At the Oct hybrid CG meeting, we decided to add back `exnref`, which was removed in 2020:
https://github.com/WebAssembly/meetings/blob/main/main/2023/CG-10.md

The new version of the proposal reflected in the explainer: https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md

While adding support for `exnref` in the current codebase which has all GC subtype hierarchies, I noticed we might need `noexn` heap type for the bottom type of `exn`. We don't have it now so I just set it to 0xff for the moment.

### CODE REVIEW

File: src/wasm/wasm-type.cpp [hunk 1]
```diff
@@ -917,6 +923,10 @@ FeatureSet Type::getFeatures() const {
                 // ref.null with just reference types.
                 feats |= FeatureSet::ReferenceTypes;
                 return;
+              case HeapType::exn:
+              case HeapType::noexn:
+                feats |= FeatureSet::ExceptionHandling;
```
Comment (tlively) [inline src/wasm/wasm-type.cpp:928]: Should we say that this requires reference types as well?

File: test/exception-handling.wast
Comment (tlively) [inline test/exception-handling.wast:1]: Would it be possible to port this test to lit, either as this PR or a separate PR? [This test](https://github.com/WebAssembly/binaryen/blob/main/test/lit/table-operations.wast) is an example where we did that port without losing any coverage.

**Review by tlively — APPROVED** [review]

File: test/exception-handling.wast
Comment (aheejin) [reply test/exception-handling.wast:1]: Will do in a follow-up!

Comment (aheejin) [reply test/exception-handling.wast:1]: #6160

**Pull request merged**